README
Vacation is the automatic mail answering program found with many Unix
implementations such as SunOS and IBM AIX.

This is the vacation port to Linux. It is derived from the vacation program
found on agate.berkeley.edu in the 386bsd directory on Nov 16, 1993. The
original program was written by Eric Allman in 1983 and is copyrighted by
the Regents of the UCB. The copyright disclaimer is valid for this
distribution as well, i.e. this version is provided `as is', and I shall not
be liable for any damage, data loss or earthquakes in the bay area.

Version 1.1 was a security-patched version. The former version 1.0 had
a bad security bug which Olaf Kirch reported to the CERT mailing list (after
giving me a hint :-). Please DO NOT LONGER USE vacation-1.0! 

As far as Copyrights are concerned, the original parts of vacation are
under the BSD copyright whereas I place my changed under the GPL (see
file COPYING in the directory).

I tried to make this port look just like the version found with SunOS 4.1.x.
There are minor differences, however, it resembles more the Sun version than
the original 386bsd source.

Differences to the 386bsd version: 
- usage of the GNU dbm database library (libdbm.a) instead of whatever
  was used in 386bsd. I don't know 386bsd well, so there was no
  chance to avoid this change. You won't bother, probably. 
- implementation of the `-j' option which suppresses checking of the
  "To:" and "Cc:" fields.
- the 386bsd version didn't replace the string "$SUBJECT" in the
  .vacation.msg file if present :-(. Well, this version does.
- the other options are named as and work just like the Sun version's,
  except:

Differences to the Sun version:
- the `-t' option only allows for an adjustment in steps of days. Not
  that it would have been to difficult to implement the whole armada of
  Sun's option, but I considered this useful enough. 
- Instead of the files .vacation.dir and .vacation.pag, respectively,
  this version uses a single file .vacation.db. The dbm_open() call used
  in the Sun and AIX versions hard-links the former files, so I dind't
  find any reason why to keep up this configuration.

There's a new option `-r' which checks if there is a "Reply-To:" header in
the incoming mail, and if there is, its entry overrides the "From:" entry.
This can be useful in those cases where users have `replyto' files for their
newsreaders for specific reasons. Plus, if someone intends to receive mail
at another location, he/she will issue "Reply-To:" as well. Feel free to
argue for or against this issue.

Unpack the package as follows

mkdir -p /usr/src		# if not already present :-)
chdir /usr/src
gzip -dc <path>/vacation-1.1.tgz | tar xpvf -

A simple `make install' (install-aout is no longer supported as of version
1.2) should install binary in /usr/bin and the manpages in /usr/man/man*,
respectively.  If you choose different directories, edit the Makefile.

There are no prerequisites as far as the compiler and the library
versions are concerned. You should be able to compile vacation with
older versions as well. However, you'll need the libdbm.a library and
the corresponding header files which should be shipped with most (if
not all) Linux distributions (at least it was with SLS 1.01). I compiled
the supplied binaries with GCC 2.6.3 (ELF). They are dynamically linked
against libc-5.4.38 and libgdbm-2.0.0. If you don't have those libraries,
either install libc-5 or recompile the binaries which should be fairly
easy.

In order to call an external shell, sendmail(8) and vi(1) correctly,
vacation #includes paths.h which should thus reflect your configuration.
Please adjust the corresponding entries in paths.h (usually found in
/usr/include).

vacation.c is still full of "#ifdef DEBUG" statements which are subject to
vanish in a later version. However, if you do some testing, the file
`/tmp/vacation.log' can be helpful. The function printd() is intended to log
into this file. Apart from that, vacation syslogs into the LOG_MAIL facility.

I apologize for not having put too many comments into the source. I felt
that for the moment, the source is very understandable itself (as you always
do when having played around with some source code for a while).

If you want to use Jeff Uphoff's "vaclook" script, simply strip everything
that isn't needed and change the permissions as appropriate. Perl5 and
dynamic loading is needed.

I'd appreciate some further testing and any useful hints concerning
vacation. Please feel free to mail me any comments.

Now maintained by Chris Samuel (chris@csamuel.org), http://www.csamuel.org/

2006/08/25


ChangeLog
2014/01/21
- Tagged and released 1.2.8.0-beta1

2013/12/26
- Added -Werror=format-security to CFLAGS for extra security checking.

2013/12/01
- Patch to handle long (folded) headers from Zdenek Havranek

2013/07/28
- Added a BuildRequires for gdbm-devel to the RPM spec file (from Richard Keech).
- Added SELinux information about Postfix on CentOS to INSTALL (from Richard Keech).
- Added the contrib script "chkvacation" to enable/disable and check vacation status (from Richard Keech).

2012/05/09
- Fixed typo in function name in comment in rfc822.c

2011/12/31
- Automated updating of version number in vacation.spec with "make gitpackage"

2011/11/23
- Automated updating of version number in vacation.h with "make gitpackage"

2011/11/10
- Added "make gitpackage" to make life easier for us maintainers!
- Also fixed up the setting of the ARCH and OS variables using the GNU make $(shell cmd) syntax, they were never getting set previously.

2011/11/08
- Install manual pages in the correct place!
- Install the German manual page in /usr/share/man/de/man1
- Support "make install DESTDIR=/foo/bar" to install to a different root.
- Simplify spec file to use make install now we support DESTDIR.

2011/11/06
- Fix location for tarballs in the vacation.spec file.
- Tagged and released Vacation 1.2.7.1.

2011/10/30
- Update version to rc2 in vacation.h, just in case, missed in rc1!
- Added Release_Checklist

2011/10/29
- HTML manpage was out of date and unmaintained, remove it and html2man and make the .man version the master.
- Added German translation of Vacation manual page courtesy of Dr. Tilmann Bubeck.
- Moved English manpage to vacation-en.man and made vacation.man a symlink to it.
- Removed "OLD" directory (untouched since 2000).
- Removed the "patches" directory which contained already applied patches.
- Tagged Vacation 1.2.7.1-rc1

2011/10/20
- Ensure orighdr is only used with DEBUG enabled, keeps GCC 4.6 quiet.
  This is a post beta2 change.

2011/10/20
- Tagged Vacation 1.2.7.1-beta2

2011/03/15
- Fix up various issues found by Coverity.

2010/11/02
- Abide by the RFC 3834 header "Auto-Submitted".

2010/06/16
- Add the Auto-Submitted: header as per RFC3834 - patch from Dr. Tilmann Bubeck.
- Applied patch from Dr. Tilmann Bubeck to stop Vacation from munging the GECOS information of users and instead pass it quoted to the MTA for it to deal with.

2008/07/13
- Added the vacation.spec file contributed by Magnus Stenman.
- Removed all references to the MAIN define from rfc822.c.
- Removed the -DMAIN compiler option from the Makefile.

2008/07/03
- Tagged Vacation 1.2.7.0

2008/06/30
- Fix last g++ warning - thanks to Daniel Pittman of Linux Users of Victoria!

2008/06/29
- Wrap _GNU_SOURCE #def in a #ifdef to suppress warning.
- Fix remaining g++ string related warnings.

2008/06/23
- Fix missing cast that stops vacation compiling with g++
- Fix various other g++ warnings due to lack of const in char * defs.

2008/06/16
- Tagged 1.2.7.0-rc2 as no more bugs reported in rc1.

2008/05/31
- Make "verbose" option code in rfc822.c depend on DEBUG being defined to supress compiler warnings.
- Fix my Id header stuffup in the ChangeLog. :-)

2008/05/19
- Fix trivial memory leak when checking for recent emails in the GDBM db.

2008/05/17
- Fix segmentation fault in strlcpy() when passed a NULL source string.
- Added KNOWN_BUGS file.
- Added Id header to rfc822.c, strlcat.c and strlcpy.c
- Tagged 1.2.7.0-rc1

2008/01/20
- Added List-Id to the headers that junkmail() searches for.
- Removed the old code for parsing email addresses, the new rfc822.c works so much better!

2008/01/19
- Mega update - reformatted everything to the GNU coding style using GNU indent
- Tidied up the ChangeLog

2008/01/17
- Added documentation for the "-h" option to specify a domain name for email replies to be from.
- Fixed bug where in DEBUG mode an unchecked fopen() would cause a crash if it failed.
- Added -g flag to use debug info by default.

2008/01/14
- Tagged 1.2.7.0 beta 5

2008/01/13
- Fixed brown paper bag bug - duplicate case statements stopped it compiling!
- Added patch from Jacek Kalinski that automatically creates a vacation database for users who do not have one.  Stops vacation dieing if they have forgotten to init the database.

2007/05/05
- Tagged 1.2.7.0 beta 4

2007/04/16
- Modified the Makefile to use uname -s to work out the OS to tell Linux from FreeBSD and set the appropriate options.

2007/04/14
- Added FreeBSD compile options to the Makefile and altered the docs.
- Thanks to Marshal Newrock for the information!

2007/01/21
- Fixed missing file for "make package"
- Simplified CFLAGS handling for clarity.
- Fixed typo in ChangeLog
- Man page updated for "-i"
- Made "-i" an alias for "-I" to initialise the database.
- Removed the -m486 compile flag for non-PPC case, seeing as the distros were doing this already, breaks backwards compatibility on AMD64 but can uncomment line in Makefile to build with -m32.

2007/01/20
- Bug with argument handling, -r expected an (unnecessary) option. Fixed!
- Added ESR's rfc822.c parser from his UNIX cookbook (and fetchmail) (MIT license)
- Modified vacation.c to use nxtaddr() from rfc822.c instead of homebrew code

vacation.c
/*
 * Copyright (c) 1988 Regents of the University of California.
 * All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Arthur David Olson of the National Cancer Institute.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the 
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *                   
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR `AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
 * EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
 * OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
 * OF SUCH DAMAGE.
 */

#ifndef lint
char copyright[] =
  "@(#) Copyright (c) 1983, 1987 Regents of the University of California.\n\
 All rights reserved.\n";
#endif /* not lint */

#ifndef lint
/*static char sccsid[] = "from: @(#)vacation.c	5.19 (Berkeley) 3/23/91";*/
static char rcsid[] __attribute__ ((unused)) =
  "$Id$";
#endif /* not lint */

/*
**  Vacation
**  Copyright (c) 1983  Eric P. Allman
**  Berkeley, California
*/

/*  ported to Linux and
**  Copyright (c) 1993-1998 Harald Milz (hm@seneca.muc.de)
*/

/*  some tiny patches recognizing/generating headers
**  Mark Seuffert (moak@pirate.de)
*/

/*  Raaah! Someone has changed the behaviour of vfork() in the kernel >= 2.2.x
**  We'll use fork instead, that works.
**  Patrik Schindler (poc@pocnet.net)
*/

/*  Aleksey Barabanov <alekseybb@mtu-net.ru>
**  Sun Feb 4 00:37:29 2001
**    1.add option -h hostname for virtual domain processing
**    2.realname in /etc/passwd with blank internals
*/

/* Patch to avoid replying to emails marked as spam by SpamAssassin
 * contributed by Roberto Piola, roberto@ilpiola.it - 2006/12/03
 */

#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif /* _GNU_SOURCE */

#include <sys/param.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <pwd.h>
#include <gdbm.h>
#include <time.h>
#include <syslog.h>
#include <errno.h>
#include <unistd.h>
#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include <string.h>
#include <paths.h>
#include "tzfile.h"
#include "vacation.h"

/* Extern definitions for getopt(3) */
extern int optind, opterr;
extern char *optarg;

/*
 *  VACATION -- return a message to the sender when on vacation.
 *
 *	This program is invoked as a message receiver.  It returns a
 *	message specified by the user to whomever sent the mail, taking
 *	care not to return a message too often to prevent "I am on
 *	vacation" loops.
 */

int
main (int argc, char **argv)
{
  struct passwd *pw;
  ALIAS *cur;
  time_t interval;
  int ch, iflag, nflag;
  char *vacation;
  char *vdomain;
  char *vusername;
  char tmpusername[ 1024 ];

  openlog ("vacation", LOG_PID, LOG_MAIL);
  opterr = iflag = nflag = rflag = 0;
  interval = -1;
  vdomain = NULL;
#ifdef _PATH_VACATION
  vacation = (char *) _PATH_VACATION;
#else
  vacation = argv[0];
#endif
  if (argc == 1)
    nflag = 1;
  while ((ch = getopt (argc, argv, "a:h:Iit:jrv")) != EOF)
    switch ((char) ch)
      {
      case 'a':		/* alias */
	if (!(cur = (ALIAS *) malloc ((u_int) sizeof (ALIAS))))
	  {
	    perror ("malloc");
	    exit (-1);
	  }
	cur->name = optarg;
	cur->next = names;
	names = cur;
	break;
      case 'h':		/* hostname */
	vdomain = optarg;
	break;
      case 'I':		/* init the database */
      case 'i':		/* init the database */
	iflag = 1;
	break;
      case 't':
	if (isdigit (*optarg))
	  {
	    interval = atol (optarg) * SECSPERDAY;	/* unit is `days' */
	    if (interval < 0)
	      usage ();
	  }
	else
	  interval = LONG_MAX;
	break;
      case 'j':		/* don't check To: and Cc: fields */
	jflag = 1;
	break;
      case 'r':		/* "Reply-To:" overrides "From:" */
	rflag = 1;
	break;
      case 'v':		/* Output vacation version number */
	printf ("Linux Vacation %s\n", VACVERS);
	exit (0);
	break;
      case '?':
      default:
	usage ();
	break;
      }
  argc -= optind;
  argv += optind;

  if (argc != 1)
    {
      if (!iflag && !nflag)
	usage ();
      if (!(pw = getpwuid (getuid ())))
	{
	  syslog (LOG_ERR, "vacation: no such user uid %u.\n", getuid ());
	  exit (1);
	}
    }
  else if (!(pw = getpwnam (*argv)))
    {
      snprintf( tmpusername, 1023, "%s", *argv );
      syslog (LOG_ERR, "vacation: no such user %s.\n", tmpusername);
      exit (1);
    }
  if (chdir (pw->pw_dir))
    {
      syslog (LOG_NOTICE, "vacation: no such directory %s.\n", pw->pw_dir);
      exit (1);
    }

  do
    {
      db = gdbm_open ((char *) VDB, 128, ((iflag || nflag) ? GDBM_NEWDB : GDBM_WRITER),
		      0644, NULL);
      if (!db && errno == ENOENT)
	{
	  db = gdbm_open ((char *) VDB, 128, GDBM_NEWDB, 0644, NULL);
	}
      if (!db && errno == EAGAIN)
	sleep ((rand () & 7) + 1);
    }
  while (!db && errno == EAGAIN);

  if (!db)
    {
      syslog (LOG_NOTICE, "vacation: %s: %s\n", VDB, strerror (errno));
      exit (1);
    }

  if (interval != -1)
    setinterval (interval);

  if (iflag)
    {
      (void) gdbm_close (db);
      exit (0);
    }

  if (nflag)
    {
      (void) gdbm_close (db);
      initialize (vacation, pw->pw_name);
      exit (0);
    }

  if (!(cur = (ALIAS *) malloc ((u_int) sizeof (ALIAS))))
    exit (1);
  cur->name = pw->pw_name;
  cur->next = names;
  names = cur;

#ifdef DEBUG
  printd ("next message");
#endif
  readheaders ();
  if (!recent ())
    {
      setreply ();
      (void) gdbm_close (db);
      if (vdomain)
	{			/* add virtual domain to username */
	  if (!(vusername = (char *) malloc (MAXLINE)))
	    {
	      perror ("malloc");
	      exit (-1);
	    }
	  (void) strlcpy (vusername, pw->pw_name, MAXLINE);
	  strlcat (vusername, "@", MAXLINE);
	  strlcat (vusername, vdomain, MAXLINE);
	  sendmessage (vusername, pw->pw_gecos);
	  free (vusername);
	}
      else
	{
	  sendmessage (pw->pw_name, pw->pw_gecos);
	}
    }
  else
    (void) gdbm_close (db);
  exit (0);
  /* NOTREACHED */
}

/*
 * readheaders --
 *	read mail headers
 */
void
readheaders (void)
{
  register ALIAS *cur;
  register char *p;
  int tome, cont;
  char buf[MAXLINE]; /* actual line */
  char buf2[MAXLINE]; /* next line */
  char uucpfrom[MAXLINE];
/*  char sender[MAXLINE]; */
/*  char domain[MAXLINE]; */

  cont = tome = 0;
  fgets (buf, sizeof (buf), stdin);
  if (*buf != '\n')
    fgets (buf2, sizeof (buf2), stdin);
  while (*buf != '\n')
   {
    while (*buf2 != '\n' && ((buf2)[0] == ' ' || (buf2)[0] == '\t') &&
	strlen (buf) + strlen (buf2) < MAXLINE) /* only to buf capacity */
      { /* it's OK for "From:" and it doesn't matter for "To:" and "Cc:" */
	(buf)[strlen (buf) - 1] = '\0'; /* remove '\n' */
	strlcat (buf, buf2, MAXLINE);
	fgets (buf2, sizeof (buf2), stdin);
      }
    switch (toupper (*buf))
      {
      case 'F':		/* "From: " or "From " */
	cont = 0;
	if (!strncasecmp (buf, "From:", 5))
	  {			/* "From:" */
	    (void) strlcpy (from, nxtaddr (buf), MAXLINE);
#ifdef DEBUG
	    snprintf (logline, MAXLINE, "From: >%s<", from);	/* Flawfinder: ignore */
	    printd (logline);
#endif
	  }
	else if (!strncasecmp (buf, "From ", 5))
	  {			/* "From " */
	    for (p = buf + 5; *p && *p != ' '; ++p);
	    *p = '\0';
	    (void) strlcpy (uucpfrom, buf + 5, MAXLINE);	/* this contains a bang path */
	    if ((p = index (uucpfrom, '\n')))
	      *p = '\0';
#ifdef DEBUG
	    snprintf (logline, MAXLINE, "From >%s<", uucpfrom);	/* Flawfinder: ignore */
	    printd (logline);
#endif
	  }
	else
	  break;		/* other "F" header */
	if (index (from, '@') > rindex (from, '.'))
	  {			/* domain addr. complete? */
	    if ((p = rindex (uucpfrom, '!')))
	      {
		strlcpy (from, p + 1, MAXLINE);
		strlcat (from, "@", MAXLINE);
		*p = '\0';
	      }
	    else
	      exit (1);		/* this should not occur */
	    if ((p = rindex (uucpfrom, '!')))
	      strlcat (from, p + 1, MAXLINE);
	    else
	      strlcat (from, uucpfrom, MAXLINE);
	  }
#ifdef DEBUG
	snprintf (logline, MAXLINE, "From: now: >%s<", from);	/* Flawfinder: ignore */
	printd (logline);
#endif
	if (junkmail ())
	  exit (0);
	break;
      case 'R':		/* "Reply-To: " */
	cont = 0;
	if (!strncasecmp (buf, "Reply-To:", 9))
	  {			/* much simpler than From: */
	    (void) strlcpy (replyto, nxtaddr (buf), MAXLINE);
#ifdef DEBUG
	    snprintf (logline, MAXLINE, "Reply-To: >%s<", replyto);	/* Flawfinder: ignore */
	    printd (logline);
#endif
	    if (junkmail ())
	      exit (0);
	  }
	break;
      case 'S':		/* "Subject" */
	cont = 0;
	if (!strncasecmp (buf, "Subject:", 8))
	  {
	    (void) strlcpy (subject, buf + 9, MAXLINE);
	    if ((p = index (subject, '\n')))
	      *p = '\0';
#ifdef DEBUG
	    snprintf (logline, MAXLINE, "Subject found: >%s<", subject);	/* Flawfinder: ignore */
	    printd (logline);
#endif
	    if (junkmail ())
	      exit (0);
	  }
	break;
      case 'P':		/* "Precedence:" */
	cont = 0;
	if (strncasecmp (buf, "Precedence", 10) ||
	    (buf[10] != ':' && buf[10] != ' ' && buf[10] != '\t'))
	  break;
	if (!(p = index (buf, ':')))
	  break;
	while (*++p && isspace (*p));
	if (!*p)
	  break;
	if (!strncasecmp (p, "junk", 4) ||
	    !strncasecmp (p, "list", 4) || !strncasecmp (p, "bulk", 4))
	  exit (0);
	break;
      case 'A':		/* "Auto-Submitted:" */
	cont = 0;
	if (strncasecmp (buf, "Auto-Submitted", 14) ||
	    (buf[14] != ':' && buf[14] != ' ' && buf[14] != '\t'))
	  break;
	if (!(p = index (buf, ':')))
	  break;
	while (*++p && isspace (*p));
	if (!*p)
	  break;
	if (strncasecmp (p, "no", 2))
	  exit (0);
	break;
      case 'C':		/* "Cc:" */
	if (strncasecmp (buf, "Cc:", 3))
	  break;
	cont = 1;
	goto findme;
      case 'T':		/* "To:" */
	if (strncasecmp (buf, "To:", 3))
	  break;
	cont = 1;
	goto findme;
      case 'X':		/* Don't reply to email marked as spam by SpamAssassin */
	if (!strncasecmp (buf, "X-Spam-Status: Yes", 18))
	  exit (0);
	break;
      default:
	if (!isspace (*buf) || !cont || tome)
	  {
	    cont = 0;
	    break;
	  }
      findme:
	for (cur = names; !tome && cur; cur = cur->next)
	  tome += nsearch (cur->name, buf);
      }
    strlcpy (buf, buf2, MAXLINE);
    if (*buf != '\n')
      fgets (buf2, sizeof (buf2), stdin);
   }
  if (!jflag && !tome)
    exit (0);
  if (!*from)
    {
      syslog (LOG_NOTICE, "vacation: no \"From:\" line.\n");
      exit (2);
    }
  if (rflag && (*replyto != 0x0))
    strlcpy (from, replyto, MAXLINE);
#ifdef DEBUG
  snprintf (logline, MAXLINE, "From: %s  Reply-To: %s\n", from, replyto);	/* Flawfinder: ignore */
  printd (logline);
#endif
}


/*
 * nsearch --
 *	do a nice, slow, search of a string for a substring.
 */
int
nsearch (register char *name, register char *str)
{
#if 0
  /* This is the old code */
  register int len;
  register char c;
  c = tolower (*name);

  for (len = strlen (name); *str; ++str)
    if (tolower (*str) == c && !strncasecmp (name, str, len))
      return (1);
  return (0);
#else
  /* New code */
  if (strcasestr (str, name))
    return (1);
  else
    return (0);
#endif
}

/*
 * junkmail --
 *	read the header and return if automagic/junk/bulk mail
 */
int
junkmail (void)
{
  static struct ignore
  {
    const char *name;
    int len;
  } ignore[] =
  {
    /* I am not sure whether this list is near to complete. Maybe a 
       separate file containing all the exceptions would be more
       handy than hardwiring this into the code. hm */
    {
    "List-Help", 9},
    {
    "List-Unsubscribe", 16},
    {
    "List-Subscribe", 14},
    {
    "List-Post", 9},
    {
    "List-Id", 7},
    {
    "List-Owner", 10},
    {
    "List-Archive", 12},
    {
    "-request", 8},
    {
    "postmaster", 10},
    {
    "uucp", 4},
    {
    "mailer-daemon", 13},
    {
    "mailer", 6},
    {
    "-relay", 6},
    {
    "-activists", 10},
    {
    "-programmers", 12},
    {
  NULL, 0},};
  register struct ignore *cur;
  register int len;
  register char *p;

  /*
   * This is mildly amusing, and I'm not positive it's right; trying
   * to find the "real" name of the sender, assuming that addresses
   * will be some variant of:
   *
   * From site!site!SENDER%site.domain%site.domain@site.domain
   */
  if (!(p = index (from, '%')))
    if (!(p = index (from, '@')))
      {
	if ((p = rindex (from, '!')))
	  ++p;
	else
	  p = from;
	for (; *p; ++p);
      }
  len = p - from;
  for (cur = ignore; cur->name; ++cur)
    if (len >= cur->len && !strncasecmp (cur->name, p - cur->len, cur->len))
      return (1);
  return (0);
}

#define	VIT	"__VACATION__INTERVAL__TIMER__"

/*
 * recent --
 *	find out if user has gotten a vacation message recently.
 *	use bcopy for machines with alignment restrictions
 */
int
recent (void)
{
  datum key, data;
  time_t then, next;

  /* get interval time */
  key.dptr = (char *) VIT;
  key.dsize = sizeof (VIT);
  data = gdbm_fetch (db, key);
  if (data.dptr == NULL)
    next = SECSPERDAY * DAYSPERWEEK;	/* default repeat time is one week */
  else
    {
      bcopy (data.dptr, &next, sizeof (next));
      /* Safely stashed in next, now we can release the GDBM pointer */
      free (data.dptr);
    }

  /* get record for this address */
  key.dptr = from;
  key.dsize = strlen (from);
  data = gdbm_fetch (db, key);
  if (data.dptr != NULL)
    {
      bcopy (data.dptr, &then, sizeof (then));
      /* Safely stashed in then, we can free the GDBM pointer now */
      free (data.dptr);
      if (next == LONG_MAX || then + next > time (NULL))
	return (1);		/* Return 1 to say we've had a recent message */
    }
  return (0);			/* Drop through to return 0 to say no recent message */
}

/*
 * setinterval --
 *	store the reply interval
 */
void
setinterval (time_t interval)
{
  datum key, data;

  key.dptr = (char *) VIT;
  key.dsize = sizeof (VIT);
  data.dptr = (char *) &interval;
  data.dsize = sizeof (interval);
  (void) gdbm_store (db, key, data, GDBM_REPLACE);
}

/*
 * setreply --
 *	store that this user knows about the vacation.
 */
void
setreply (void)
{
  datum key, data;
  time_t now;

  key.dptr = from;
  key.dsize = strlen (from);
  (void) time (&now);
  data.dptr = (char *) &now;
  data.dsize = sizeof (now);
  (void) gdbm_store (db, key, data, GDBM_REPLACE);
}

/*
 * sendmessage --
 *	exec sendmail to send the vacation file to sender
 */
void
sendmessage (char *myname, char *myrealname)
{
  FILE *mfp, *sfp;
  int i;
  int pvect[2];
  char buf[MAXLINE];
  char line[MAXLINE];

#ifdef DEBUG
  snprintf (logline, MAXLINE, "sendmessage: using realname >%s<", myrealname);	/* Flawfinder: ignore */
  printd (logline);
#endif

  mfp = fopen (VMSG, "r");
  if (mfp == NULL)
    {
      syslog (LOG_NOTICE, "vacation: no ~%s/%s file.\n", myname, VMSG);
      exit (1);
    }
  if (pipe (pvect) < 0)
    {
      syslog (LOG_ERR, "vacation: pipe: %s", strerror (errno));
      exit (1);
    }
  i = fork ();
  if (i < 0)
    {
      syslog (LOG_ERR, "vacation: fork: %s", strerror (errno));
      exit (1);
    }
  if (i == 0)
    {
      dup2 (pvect[0], 0);
      close (pvect[0]);
      close (pvect[1]);
      fclose (mfp);
      execl (_PATH_SENDMAIL, "sendmail", "-f", myname, "-F", myrealname, "--", from,
	     NULL);
      syslog (LOG_ERR, "vacation: can't exec %s: %s", _PATH_SENDMAIL,
	      strerror (errno));
      exit (1);
    }
  close (pvect[0]);
  sfp = fdopen (pvect[1], "w");
  fprintf (sfp, "Precedence: bulk\n");
  fprintf (sfp, "Auto-Submitted: auto-replied\n" ); /* RFC-3834 */
  fprintf (sfp, "User-Agent: %s http://vacation.sourceforge.net\n", VACVERS);
  fprintf (sfp, "To: %s\n", from);
  while (fgets (buf, sizeof buf, mfp))
    {
      char *sp, *fromp, *subjp, *nextp;

      subjp = strstr (buf, "$SUBJECT");
#ifdef DEBUG
      if (subjp)
	{
	  snprintf (logline, MAXLINE, "sendmessage: found $SUBJECT %s\n", subjp);	/* Flawfinder: ignore */
	  printd (logline);
	}
#endif
      fromp = strstr (buf, "$FROM");
      for (sp = buf; subjp || fromp; sp = nextp)
	{
	  if (subjp && !(fromp && fromp < subjp))
	    {
	      *subjp = '\0';
	      nextp = subjp + 8;
	      snprintf (line, MAXLINE, "%s%s", sp, subject);	/* Flawfinder: ignore */
	      subjp = NULL;
	    }
	  else
	    {
	      *fromp = '\0';
	      nextp = fromp + 5;
	      snprintf (line, MAXLINE, "%s%s", sp, from);	/* Flawfinder: ignore */
	      fromp = NULL;
	    }
	  fputs (line, sfp);
	}
      fputs (sp, sfp);
    }
  fclose (mfp);
  fclose (sfp);
}

void
usage (void)
{
  puts ("usage: vacation [ -I ]");
  puts
    ("or:    vacation [ -j ] [ -a alias ] [ -h hostname ] [ -tN ] [ -r ] login");
  exit (1);
}


void
initialize (char *path, char *myname)
{
  char *editor;
  char ebuf[PATH_MAX];
  int  c;
  FILE *message;
  FILE *oldmessage;
  FILE *forward;
  FILE *oldforward;

  if (((forward = fopen (FWD, "r")) == NULL))
    {
      if (((forward = fopen (FWD, "w")) == NULL))
	{
	  fprintf (stderr, "vacation: cannot open %s for reading\n", FWD);
	  exit (1);
	}
    }
  else
    {
      if (((oldforward = fopen (OLDFWD, "w")) == NULL))
	{
	  fprintf (stderr, "vacation: cannot open %s for writing\n", OLDFWD);
	  exit (1);
	}
      while (((c = fgetc (forward)) != EOF))
	{
	  fprintf (oldforward, "%c", ( unsigned char ) c);
	  chmod (OLDFWD, 00600);
	}
      fclose (oldforward);
      if (((forward = freopen (FWD, "w", forward)) == NULL))
	{
	  fprintf (stderr, "vacation: cannot open %s for writing\n", FWD);
	  exit (1);
	}
    }
#ifndef COURIER
  fprintf (forward, "\\%s, \"|%s %s\"\n", myname, path, myname);
#else
  fprintf (forward, "./Maildir\n|%s %s\n", path, myname);
#endif

  fclose (forward);
  chmod (FWD, 00600);

  if (((editor = getenv ("VISUAL")) == NULL))
    if (((editor = getenv ("EDITOR")) == NULL))
      editor = (char *) _PATH_VI;
#ifdef DEBUG
  snprintf (logline, MAXLINE, "editor: %s\n", editor);	/* Flawfinder: ignore */
  printd (logline);
#endif

  if (((message = fopen (VMSG, "r")) == NULL))
    {
      if (((message = fopen (VMSG, "w")) == NULL))
	{
	  fprintf (stderr, "vacation: cannot open %s\n", VMSG);
	  exit (1);
	}
      fprintf (message, "Subject: away from my mail\n\n");
      fprintf (message, "I will not be reading my mail for a while.\n");
      fprintf (message, "Your mail concerning \"$SUBJECT\"\n");
      fprintf (message, "will be read when I'm back.\n");
    }
  else
    {
      if (((oldmessage = fopen (OLDVMSG, "w")) == NULL))
	{
	  fprintf (stderr, "vacation: cannot open %s\n", OLDVMSG);
	  exit (1);
	}
      while (((c = fgetc (message)) != EOF))
	{
	  fprintf (oldmessage, "%c", ( unsigned char ) c);
	}
      fclose (oldmessage);
    }
  fclose (message);
  snprintf (ebuf, PATH_MAX, "%s %s", editor, VMSG);	/* Flawfinder: ignore */
#ifdef DEBUG
  snprintf (logline, MAXLINE, "calling editor with \"%s\"\n", ebuf);	/* Flawfinder: ignore */
  printd (logline);
#endif
  system (ebuf);
  exit (0);
}


#ifdef DEBUG
void
printd (char *message)
{
  FILE *file;
  time_t now;
  char *timestr;
  char *p;

  time (&now);
  timestr = ctime (&now);
  if ((p = index (timestr, '\n')))
    *p = '\0';
  file = fopen ("/tmp/vacation.log", "a");
  if (!file)
    {
      syslog (LOG_ERR,
	      "vacation: failed to open /tmp/vacation.log (errno=%d).\n",
	      errno);
      return;
    }
  fprintf (file, "%s %s\n", timestr, message);
  fclose (file);
}
#endif
